tidy3d/components/grid/mesher.py

```diff
@@ -86,94 +90,158 @@ def parse_structures(
+
+            # Handle insertion of the current structure bounds in the intervals
+            intervals = self.insert_bbox(intervals, str_ind, bbox, bbox_contained_2d, min_step)
 
         # Truncate intervals to domain bounds
-        b_array = np.array(interval_coords)
-        in_domain = np.argwhere((b_array >= domain_bounds[0]) * (b_array <= domain_bounds[1]))
-        interval_coords = interval_coords[in_domain]
-        interval_structs = [interval_structs[int(i)] for i in in_domain if i < b_array.size - 1]
-
-        # Remove intervals that are smaller than the absolute smallest min_step
-        min_step = np.amin(structure_steps)
-        coords_filter = [interval_coords[0]]
-        structs_filter = []
-        for coord_ind, coord in enumerate(interval_coords[1:]):
-            if coord - coords_filter[-1] > min_step:
-                coords_filter.append(coord)
-                structs_filter.append(interval_structs[coord_ind])
-        interval_coords = np.array(coords_filter)
-        interval_structs = structs_filter
+        coords = np.array(intervals["coords"])
+        num_ints = len(intervals["structs"])
+        in_domain = np.argwhere((coords >= domain_bounds[0]) * (coords <= domain_bounds[1]))
+        intervals["coords"] = [intervals["coords"][int(i)] for i in in_domain]
+        intervals["structs"] = [intervals["structs"][int(i)] for i in in_domain if i < num_ints]
 
         # Compute the maximum allowed step size in each interval
         max_steps = []
-        for coord_ind, _ in enumerate(interval_coords[:-1]):
-            # Structure indexes inside current intervall in order of latter structures come first
-            struct_list = interval_structs[coord_ind]
-            struct_list_filter = []
-            # Handle containment
-            for ind, struct_ind in enumerate(struct_list[::-1]):
-                if ind >= len(struct_list):
-                    # This can happen because we modify struct_list in the loop
-                    break
-                # Add current structure to filtered list
-                struct_list_filter.append(struct_ind)
-                # Remove all structures that current structure overrides
-                contains = struct_contains[struct_ind]
-                struct_list = [ind for ind in struct_list if ind not in contains]
-
+        for coord_ind, _ in enumerate(intervals["coords"][:-1]):
             # Define the max step as the minimum over all medium steps of media in this interval
-            max_step = np.amin(structure_steps[struct_list_filter])
+            max_step = np.amin(structure_steps[intervals["structs"][coord_ind]])
             max_steps.append(float(max_step))
 
-        return interval_coords, np.array(max_steps)
+        # Re-evaluate the absolute smallest min_step and remove intervals that are smaller than that
+        intervals["coords"], max_steps = self.filter_min_step(intervals["coords"], max_steps)
```

Comment:
Will any intervals actually be filtered here? If do, will something weird occur at Line201-202? E.g. some coords are too close here so that more than the closest neighbor comparison is needed.